File: python/src/scmarkeragent/viewer.py
"""Build a portable, reference-free viewer for one production result.

The viewer's CSS, JS and HTML skeleton live in `resources/static/viewer/`, SHARED with
the R arm's `rflow/viewer.R`: both arms splice the same bytes around the same payload,
which is what makes the delivered `index.html` identical no matter which arm wrote it.
"""

from __future__ import annotations

import csv
import html
import json
import math
import shutil
import zipfile
from datetime import UTC, datetime
from pathlib import Path

_STATIC = Path(__file__).resolve().parent / "resources" / "static" / "viewer"
CSS = (_STATIC / "viewer.css").read_text(encoding="utf-8")
JS = (_STATIC / "viewer.js").read_text(encoding="utf-8")
_TEMPLATE = (_STATIC / "index_template.html").read_text(encoding="utf-8")

def _read_csv(path: Path) -> list[dict]:
    with path.open(encoding="utf-8", newline="") as handle:
        return list(csv.DictReader(handle))


def _read_jsonl(path: Path) -> list[dict]:
    return [
        json.loads(line)
        for line in path.read_text(encoding="utf-8").splitlines()
        if line
    ]


def _safe_json(value) -> str:
    def clean(item):
        if isinstance(item, dict):
            return {str(key): clean(entry) for key, entry in item.items()}
        if isinstance(item, list):
            return [clean(entry) for entry in item]
        if isinstance(item, float) and (math.isnan(item) or math.isinf(item)):
            return None
        return item

    return (
        json.dumps(
            clean(value), ensure_ascii=False, separators=(",", ":"), allow_nan=False
        )
        .replace("</", "<\\/")
        .replace("<!--", "<\\!--")
    )


def _copy_viewer_files(result_dir: Path, viewer_dir: Path) -> list[str]:
    assets = viewer_dir / "assets" / "figures"
    downloads = viewer_dir / "downloads"
    assets.mkdir(parents=True)
    downloads.mkdir(parents=True)
    copied = []
    figure_names = (
        "dotplot_celltype_markers.png",
        "dotplot_celltype_markers_by_celltype.png",
        "dotplot_celltype_markers_publication_support.png",
        "umap_celltype_labeled.png",
        "umap_cluster.png",
    )
    for name in figure_names:
        source = result_dir / "figures" / name
        if source.is_file():
            target = assets / name
            shutil.copy2(source, target)
            copied.append(str(target.relative_to(viewer_dir)))
    for name in (
        "cluster_summary.csv",
        "marker_evidence.csv",
        "cluster_evidence.jsonl",
        "markers_all_by_cluster.csv",
        "markers_significant_by_cluster.csv",
    ):
        source = result_dir / name
        if not source.is_file():
            continue
        target = downloads / name
        shutil.copy2(source, target)
        copied.append(str(target.relative_to(viewer_dir)))
    for source in sorted((result_dir / "figures").glob("*.pdf")):
        target = downloads / source.name
        shutil.copy2(source, target)
        copied.append(str(target.relative_to(viewer_dir)))
    return copied


def _figure_html(viewer_dir: Path) -> tuple[str, str]:
    figures = [
        ("identity-markers", "Identity markers", "dotplot_celltype_markers.png"),
        (
            "identity-markers-celltype",
            "Identity markers by cell type",
            "dotplot_celltype_markers_by_celltype.png",
        ),
        (
            "identity-markers-support",
            "Publication support",
            "dotplot_celltype_markers_publication_support.png",
        ),
        ("cell-types", "Cell types", "umap_celltype_labeled.png"),
        ("clusters", "Clusters", "umap_cluster.png"),
    ]
    available = [
        item
        for item in figures
        if (viewer_dir / "assets" / "figures" / item[2]).is_file()
    ]
    tabs = "".join(
        f'<button class="tab figure-tab{" active" if index == 0 else ""}" '
        f'data-figure="{identifier}">{html.escape(label)}</button>'
        for index, (identifier, label, _) in enumerate(available)
    )
    images = "".join(
        f'<img id="{identifier}" class="{"active" if index == 0 else ""}" '
        f'src="assets/figures/{html.escape(filename)}" alt="{html.escape(label)}">'
        for index, (identifier, label, filename) in enumerate(available)
    )
    return tabs, images


def _build_html(tag: str, result_dir: Path, viewer_dir: Path) -> str:
    clusters = _read_csv(result_dir / "cluster_summary.csv")
    markers = _read_csv(result_dir / "marker_evidence.csv")
    evidence = _read_jsonl(result_dir / "cluster_evidence.jsonl")
    total_cells = sum(int(row["n_cells"]) for row in clusters)
    resolutions = sorted(
        {row["resolution_status"] for row in clusters if row["resolution_status"]}
    )
    options = '<option value="">All resolution states</option>' + "".join(
        f'<option value="{html.escape(value)}">{html.escape(value)}</option>'
        for value in resolutions
    )
    figure_tabs, figure_images = _figure_html(viewer_dir)
    data = _safe_json({"clusters": clusters, "markers": markers, "evidence": evidence})
    return f"""<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>{html.escape(tag)} — scMarkerAgent</title>
<style>{CSS}</style>
</head>
<body>
<header class="topbar"><div class="topbar-inner">
  <div class="brand">scMarkerAgent</div>
  <div class="contract">Reference-free production result · one dataset</div>
</div></header>
<main>
  <h1>{html.escape(tag)}</h1>
  <p class="lead">Marker-grounded cluster annotations with measured, LLM-validated,
  source-backed cell-type identity markers.</p>
  <div class="notice"><strong>Reference-free production output.</strong>
  No author annotation, purity, reference metric, or secondary label is read or displayed.</div>
  <div class="cards">
    <div class="card"><div class="metric-label">Cells</div><div class="metric-value">{total_cells:,}</div><div class="metric-note">after production QC</div></div>
    <div class="card"><div class="metric-label">Clusters</div><div class="metric-value">{len(clusters)}</div><div class="metric-note">one reported annotation per cluster</div></div>
    <div class="card"><div class="metric-label">Identity markers</div><div class="metric-value">{len(markers):,}</div><div class="metric-note">LLM-validated and DB-cited</div></div>
  </div>
  <div class="downloads">
    <a class="button" href="downloads/cluster_summary.csv">Cluster summary</a>
    <a class="button" href="downloads/marker_evidence.csv">Identity marker evidence</a>
    <a class="button" href="downloads/markers_all_by_cluster.csv">All markers per cluster</a>
    <a class="button" href="downloads/markers_significant_by_cluster.csv">Significant markers per cluster</a>
    <a class="button" href="downloads/cluster_evidence.jsonl">Structured evidence</a>
  </div>
  <section class="panel"><h2>Figures</h2><div class="tabs">{figure_tabs}</div>
    <div class="figure-stage">{figure_images}</div>
  </section>
  <section class="panel"><h2>Cluster explorer</h2>
    <div class="toolbar">
      <input id="cluster-search" type="search" placeholder="Search cluster, annotation, or marker">
      <select id="resolution-filter">{options}</select>
      <span id="cluster-count" class="count"></span>
    </div>
    <div class="table-wrap"><table><thead><tr>
      <th class="num"><button data-sort="cluster_id">Cluster</button></th>
      <th class="num"><button data-sort="n_cells">Cells</button></th>
      <th><button data-sort="cell_type_annotation">Cell type</button></th>
      <th><button data-sort="cell_subtype_annotation">Cell subtype</button></th>
      <th><button data-sort="cell_ontology">Cell Ontology</button></th>
      <th><button data-sort="annotation_confidence">Confidence</button></th>
      <th><button data-sort="resolution_status">Resolution</button></th>
      <th><button data-sort="annotation_source">Decided by</button></th>
    </tr></thead><tbody id="cluster-body"></tbody></table></div>
  </section>
  <section id="cluster-detail" class="panel detail"><div class="empty">Select a cluster.</div></section>
  <p class="footer">Confidence is the high / medium / low judgement the annotating agent made
  from the supplied evidence. Every retrieved candidate, its independent review, and the three
  relative scores behind the retrieval order are kept in the structured evidence file. The
  Cell Ontology column is displayed as curated and is never used to decide, merge or rank an
  annotation; N/A means the curated free-text cell type carries no ontology id.</p>
</main>
<script id="result-data" type="application/json">{data}</script>
<script>{JS}</script>
</body>
</html>
"""


def _write_zip(viewer_dir: Path, output: Path) -> None:
    with zipfile.ZipFile(output, "w", compression=zipfile.ZIP_DEFLATED) as archive:
        for path in sorted(viewer_dir.rglob("*")):
            if path.is_file():
                archive.write(path, path.relative_to(viewer_dir).as_posix())


def build_dataset_viewer(tag: str, outdir: str | Path) -> dict[str, str]:
    result_dir = Path(outdir)
    viewer_dir = result_dir / "viewer"
    if viewer_dir.exists():
        shutil.rmtree(viewer_dir)
    viewer_dir.mkdir(parents=True)
    copied = _copy_viewer_files(result_dir, viewer_dir)
    html_path = viewer_dir / "index.html"
    html_path.write_text(
        _build_html(tag, result_dir, viewer_dir),
        encoding="utf-8",
    )
    copied.append("index.html")
    manifest = {
        "schema_version": "scmarkeragent-dataset-viewer-v1",
        "dataset": tag,
        "reference_free": True,
        "generated_at_utc": datetime.now(UTC).isoformat(),
        "entrypoint": "index.html",
        "files": sorted(copied),
    }
    manifest_path = viewer_dir / "viewer_manifest.json"
    manifest_path.write_text(
        json.dumps(manifest, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    zip_path = result_dir / f"{tag}_results.zip"
    if zip_path.exists():
        zip_path.unlink()
    _write_zip(viewer_dir, zip_path)
    return {
        "viewer": str(html_path.resolve()),
        "viewer_zip": str(zip_path.resolve()),
    }

File: R/inst/scmarkeragent/rflow/viewer.R
# =============================================================================
# viewer.R -- R-native builder of the portable, reference-free dataset viewer.
#
# Every arm reads the SAME CSS / JS / HTML skeleton from resources/static/viewer/ and
# splices the same payload into it, so `viewer/index.html` is identical down to the byte
# no matter which arm wrote it -- which is exactly what the parity comparator asserts
# for .html files.
#
# Reads only delivered files (tables, JSONL, figures); never re-decides anything.
# =============================================================================

suppressPackageStartupMessages({
  library(data.table)
})

.viewer_script_dir <- function() {
  frame_files <- vapply(
    sys.frames(),
    function(frame) {
      value <- frame$ofile
      if (is.null(value)) "" else as.character(value)[1]
    },
    ""
  )
  frame_files <- frame_files[nzchar(frame_files)]
  if (length(frame_files)) {
    return(dirname(normalizePath(frame_files[length(frame_files)])))
  }
  args <- commandArgs(trailingOnly = FALSE)
  match <- grep("^--file=", args, value = TRUE)
  if (length(match)) {
    return(dirname(normalizePath(sub("^--file=", "", match[1]))))
  }
  normalizePath(getwd())
}
.VIEWER_SELF <- .viewer_script_dir()
.VIEWER_STATIC <- file.path(
  dirname(.VIEWER_SELF), "resources", "static", "viewer"
)

.viewer_read_static <- function(name) {
  path <- file.path(.VIEWER_STATIC, name)
  if (!file.exists(path)) {
    stop(sprintf("missing shared viewer asset: %s", path), call. = FALSE)
  }
  readChar(path, file.info(path)$size, useBytes = TRUE)
}

#' Replace every occurrence of `token` with `value`, both taken literally.
#'
#' gsub() interprets backslashes in its replacement even under fixed = TRUE, and the
#' payload is full of them; splitting sidesteps the whole escape question.
.viewer_splice <- function(text, token, value) {
  parts <- strsplit(text, token, fixed = TRUE)[[1]]
  paste(parts, collapse = value)
}

#' Escape the five HTML entities, ampersand first.
.viewer_html_escape <- function(text) {
  text <- gsub("&", "&amp;", text, fixed = TRUE)
  text <- gsub("<", "&lt;", text, fixed = TRUE)
  text <- gsub(">", "&gt;", text, fixed = TRUE)
  text <- gsub("\"", "&quot;", text, fixed = TRUE)
  gsub("'", "&#x27;", text, fixed = TRUE)
}

#' JSON string escaping: quote, backslash and control characters; everything else raw
#' UTF-8. Kept local so the viewer stays sourceable on its own.
.viewer_json_escape <- function(text) {
  text <- enc2utf8(as.character(text))
  text <- gsub("\\", "\\\\", text, fixed = TRUE)
  text <- gsub("\"", "\\\"", text, fixed = TRUE)
  text <- gsub("\n", "\\n", text, fixed = TRUE)
  text <- gsub("\r", "\\r", text, fixed = TRUE)
  text <- gsub("\t", "\\t", text, fixed = TRUE)
  if (grepl("[\x01-\x1f]", text, useBytes = TRUE)) {
    for (code in c(1:7, 11L, 14:31)) {
      text <- gsub(intToUtf8(code), sprintf("\\u%04x", code), text, fixed = TRUE)
    }
    text <- gsub("\b", "\\b", text, fixed = TRUE)
    text <- gsub("\f", "\\f", text, fixed = TRUE)
  }
  paste0("\"", text, "\"")
}

#' An integer written with thousands separators.
.viewer_thousands <- function(n) {
  format(as.integer(n), big.mark = ",", scientific = FALSE, trim = TRUE)
}

#' A delivered CSV as an array of JSON objects, one per row, keys in column order,
#' every value the exact string the file carries.
.viewer_table_json <- function(path) {
  table <- fread(
    path,
    colClasses = "character", na.strings = NULL, encoding = "UTF-8",
    keepLeadingZeros = TRUE
  )
  if (!nrow(table)) {
    return("[]")
  }
  columns <- names(table)
  keys <- vapply(columns, .viewer_json_escape, "")
  rows <- vapply(
    seq_len(nrow(table)),
    function(i) {
      values <- vapply(
        columns,
        function(column) .viewer_json_escape(table[[column]][i]),
        ""
      )
      paste0("{", paste(paste0(keys, ":", values), collapse = ","), "}")
    },
    ""
  )
  paste0("[", paste(rows, collapse = ","), "]")
}

# The five figure panels the viewer offers, in display order; absent files are skipped.
.VIEWER_FIGURES <- list(
  list(id = "identity-markers", label = "Identity markers",
       file = "dotplot_celltype_markers.png"),
  list(id = "identity-markers-celltype", label = "Identity markers by cell type",
       file = "dotplot_celltype_markers_by_celltype.png"),
  list(id = "identity-markers-support", label = "Publication support",
       file = "dotplot_celltype_markers_publication_support.png"),
  list(id = "cell-types", label = "Cell types", file = "umap_celltype_labeled.png"),
  list(id = "clusters", label = "Clusters", file = "umap_cluster.png")
)

.VIEWER_DOWNLOAD_TABLES <- c(
  "cluster_summary.csv", "marker_evidence.csv", "cluster_evidence.jsonl",
  "markers_all_by_cluster.csv", "markers_significant_by_cluster.csv"
)

.viewer_copy_files <- function(result_dir, viewer_dir) {
  assets <- file.path(viewer_dir, "assets", "figures")
  downloads <- file.path(viewer_dir, "downloads")
  dir.create(assets, recursive = TRUE)
  dir.create(downloads, recursive = TRUE)
  copied <- character(0)
  for (figure in .VIEWER_FIGURES) {
    source_path <- file.path(result_dir, "figures", figure$file)
    if (file.exists(source_path)) {
      file.copy(source_path, file.path(assets, figure$file), copy.date = TRUE)
      copied <- c(copied, file.path("assets", "figures", figure$file))
    }
  }
  for (name in .VIEWER_DOWNLOAD_TABLES) {
    source_path <- file.path(result_dir, name)
    if (!file.exists(source_path)) next
    file.copy(source_path, file.path(downloads, name), copy.date = TRUE)
    copied <- c(copied, file.path("downloads", name))
  }
  pdfs <- sort(list.files(file.path(result_dir, "figures"), pattern = "\\.pdf$"))
  for (name in pdfs) {
    file.copy(
      file.path(result_dir, "figures", name),
      file.path(downloads, name),
      copy.date = TRUE
    )
    copied <- c(copied, file.path("downloads", name))
  }
  copied
}

.viewer_figure_html <- function(viewer_dir) {
  available <- Filter(
    function(figure) {
      file.exists(file.path(viewer_dir, "assets", "figures", figure$file))
    },
    .VIEWER_FIGURES
  )
  tabs <- vapply(
    seq_along(available),
    function(index) {
      figure <- available[[index]]
      sprintf(
        '<button class="tab figure-tab%s" data-figure="%s">%s</button>',
        if (index == 1L) " active" else "",
        figure$id,
        .viewer_html_escape(figure$label)
      )
    },
    ""
  )
  images <- vapply(
    seq_along(available),
    function(index) {
      figure <- available[[index]]
      sprintf(
        '<img id="%s" class="%s" src="assets/figures/%s" alt="%s">',
        figure$id,
        if (index == 1L) "active" else "",
        .viewer_html_escape(figure$file),
        .viewer_html_escape(figure$label)
      )
    },
    ""
  )
  list(tabs = paste(tabs, collapse = ""), images = paste(images, collapse = ""))
}

.viewer_build_html <- function(tag, result_dir, viewer_dir) {
  clusters <- fread(
    file.path(result_dir, "cluster_summary.csv"),
    colClasses = "character", na.strings = NULL, encoding = "UTF-8"
  )
  total_cells <- sum(as.integer(clusters$n_cells))
  resolutions <- sort(
    unique(clusters$resolution_status[nzchar(clusters$resolution_status)]),
    method = "radix"
  )
  options <- paste0(
    '<option value="">All resolution states</option>',
    paste(
      vapply(
        resolutions,
        function(value) {
          escaped <- .viewer_html_escape(value)
          sprintf('<option value="%s">%s</option>', escaped, escaped)
        },
        ""
      ),
      collapse = ""
    )
  )
  figures <- .viewer_figure_html(viewer_dir)

  # The payload: the two tables re-serialized cell for cell, and the audit sidecar's
  # own lines spliced in verbatim -- parsing and re-serializing a line this package
  # wrote returns that line, so embedding the bytes equals re-serializing them.
  evidence_lines <- readLines(
    file.path(result_dir, "cluster_evidence.jsonl"),
    encoding = "UTF-8", warn = FALSE
  )
  evidence_lines <- evidence_lines[nzchar(trimws(evidence_lines))]
  data <- paste0(
    '{"clusters":', .viewer_table_json(file.path(result_dir, "cluster_summary.csv")),
    ',"markers":', .viewer_table_json(file.path(result_dir, "marker_evidence.csv")),
    ',"evidence":[', paste(evidence_lines, collapse = ","), "]}"
  )
  data <- gsub("</", "<\\\\/", data, fixed = TRUE)
  data <- gsub("<!--", "<\\\\!--", data, fixed = TRUE)

  page <- .viewer_read_static("index_template.html")
  replacements <- list(
    "__SCMA_TAG__" = .viewer_html_escape(tag),
    "__SCMA_CSS__" = .viewer_read_static("viewer.css"),
    "__SCMA_JS__" = .viewer_read_static("viewer.js"),
    "__SCMA_TOTAL_CELLS__" = .viewer_thousands(total_cells),
    "__SCMA_N_CLUSTERS__" = as.character(nrow(clusters)),
    "__SCMA_N_MARKERS__" = .viewer_thousands(
      nrow(fread(
        file.path(result_dir, "marker_evidence.csv"),
        colClasses = "character", na.strings = NULL, encoding = "UTF-8"
      ))
    ),
    "__SCMA_RESOLUTION_OPTIONS__" = options,
    "__SCMA_FIGURE_TABS__" = figures$tabs,
    "__SCMA_FIGURE_IMAGES__" = figures$images,
    "__SCMA_DATA__" = data
  )
  for (token in names(replacements)) {
    page <- .viewer_splice(page, token, replacements[[token]])
  }
  page
}

.viewer_write_zip <- function(viewer_dir, output) {
  if (!requireNamespace("zip", quietly = TRUE)) {
    stop("the 'zip' package is required to write the results archive", call. = FALSE)
  }
  members <- sort(list.files(viewer_dir, recursive = TRUE, all.files = TRUE))
  if (file.exists(output)) unlink(output)
  # mirror mode stores each member under its path relative to root -- cherry-pick
  # would flatten them to basenames, which is a different archive, not a different
  # spelling of this one.
  zip::zip(
    zipfile = output, files = members, root = viewer_dir,
    mode = "mirror", include_directories = FALSE
  )
  invisible(output)
}

#' Build the viewer directory, its manifest and the delivered ZIP for one dataset --
#' the counterpart of `viewer.build_dataset_viewer`.
scma_build_dataset_viewer <- function(tag, outdir) {
  result_dir <- normalizePath(outdir)
  viewer_dir <- file.path(result_dir, "viewer")
  if (dir.exists(viewer_dir)) {
    unlink(viewer_dir, recursive = TRUE)
  }
  dir.create(viewer_dir, recursive = TRUE)
  copied <- .viewer_copy_files(result_dir, viewer_dir)

  html_path <- file.path(viewer_dir, "index.html")
  connection <- file(html_path, open = "wb")
  writeChar(
    enc2utf8(.viewer_build_html(tag, result_dir, viewer_dir)),
    connection,
    eos = NULL, useBytes = TRUE
  )
  close(connection)
  copied <- c(copied, "index.html")

  manifest <- list(
    schema_version = "scmarkeragent-dataset-viewer-v1",
    dataset = tag,
    reference_free = TRUE,
    generated_at_utc = strftime(
      as.POSIXlt(Sys.time(), tz = "UTC"), "%Y-%m-%dT%H:%M:%OS6+00:00"
    ),
    entrypoint = "index.html",
    files = as.list(sort(copied, method = "radix"))
  )
  manifest_path <- file.path(viewer_dir, "viewer_manifest.json")
  manifest_json <- jsonlite::toJSON(
    manifest,
    auto_unbox = TRUE, pretty = 2, null = "null"
  )
  manifest_connection <- file(manifest_path, open = "wb")
  writeChar(
    enc2utf8(paste0(as.character(manifest_json), "\n")),
    manifest_connection,
    eos = NULL, useBytes = TRUE
  )
  close(manifest_connection)

  zip_path <- file.path(result_dir, sprintf("%s_results.zip", tag))
  .viewer_write_zip(viewer_dir, zip_path)
  list(
    viewer = normalizePath(html_path),
    viewer_zip = normalizePath(zip_path)
  )
}
